RFC-161 - Technical Direction in Publishing 2023 (#161)
alphagov/govuk-rfcs: rfc-161-technical-direction-in-publishing-2023.md

# Technical Direction in Publishing 2023

## Summary

This RFC documents the expected future direction of applications in the Publishing area of GOV.UK and the technical principles that are adopted by teams working in Publishing as of Summer 2023. While the information in this can be, and is encouraged, to be used by other parts of GOV.UK we have only sought product level agreement within Publishing.

## Problem

> This RFC aims to address a number of cases where there is information that is shared conversationally but a lack of documentation of it. It reflects that the future app direction in Publishing is no longer consistent with [RFC 95 - long term future of applications](https://github.com/alphagov/govuk-rfcs/blob/main/rfc-095-long-term-future-apps.md) and that Publishing has gone through and continues to go through frequent changes in people.

Author: Yup, I plan to do that on the journey to merging this

content-store | API for content on GOV.UK | No change expected | Publishing | Short term: replace Mongo 2.6 with PostgreSQL<br>Long term: evaluate whether it would be preferable to implement Content Store functionality in Publishing API and retire Content Store
content-tagger | Tool to tag content and manage the taxonomy on GOV.UK | No change expected | Publishing | Medium term: remove functionality related to Topic Taxonomy rollout<br>Long term: future unclear but will likely depend on product direction/investment in Topic and World Taxonomies
hmrc-manuals-api | API for HMRC to publish manuals to GOV.UK | No change expected | Publishing | Long term: undesirable to have an app for one document type and one department; we should decide whether we are prepared to provide API publishing functionality to departments and, if so, provide it in a reusable way that would eventually replace this
info-frontend | Serves /info pages to display user needs and performance data about a page on GOV.UK | No change expected | User Experience | Short term: retire due to being rendered mostly obsolete by Performance Platform decommissioning and planned Maslow retirement

> manuals-publisher | Publishes manuals on GOV.UK | Retire in favour of content-publisher | Publishing | Short term: deliver quality of life improvementsLong term: consider whether app can be migrated into Whitehall, should Whitehall have support for hierarchical content structures.

Reviewer: ```suggestion
manuals-publisher | Publishes manuals on GOV.UK | Retire in favour of content-publisher | Publishing | Short term: deliver quality of life improvements<br>Long term: consider whether app can be migrated into Whitehall, should Whitehall have support for hierarchical content structures.
```

publishing-api | API to publish content on GOV.UK | No change expected | Publishing | Medium term: iterate to meet emerging business needs, such as embedded content<br>Long term: potential consolidation destination for Asset Manager, Content Store and Router API
router-api | API for updating the routes used by the router on GOV.UK | Merge into either content store or publishing-api | User Experience | Long term: establish an alternative way for Router to load routes from Publishing API or Content Store and then retire
service-manual-publisher | Publishes the Service Manual on GOV.UK | Custom application that could long term be replaced by content-publisher | Publishing | Long term: candidate for consolidation into Whitehall
short-url-manager | Tool to request, approve and create short URL redirects on GOV.UK | No change expected | Publishing | No change expected

> signon | Single sign-on service for GOV.UK | No change expected | Publishing | Short term: invest in security and modernisation updates<br>Medium/long term: explore whether off-the-shelf software or third party integrations could replace this software or reduce it’s responsibilities

Reviewer: ```suggestion
signon | Single sign-on service for GOV.UK | No change expected | Publishing | Short term: invest in security and modernisation updates<br>Medium/long term: explore whether off-the-shelf software or third party integrations could replace this software or reduce its responsibilities
```

Replacing whole applications by rebuilding their features in a new application should only be considered when it can be completed in a short timespan or if the underlying framework and/or programming language is end-of-life. Rebuilding of components/modules within software applications is reasonable when it can be completed in a sufficiently short time frame that a team can guarantee its completion and the completion of any migratory work. 

#### Only build new apps in exceptional circumstances 

We have [lots of apps](https://docs.publishing.service.gov.uk/apps.html), far more than we can realistically maintain and iterate - we don’t want to add to this (indeed, we’d like to decrease it). We aspire to have our apps meet a broad section of needs for the majority of publishers, rather than being bespoke functionality for niche scenarios. We’ve learned many times that it’s quick to create a new app, hard to justify continuously financing them and harder still, once they have some usage, to retire or consolidate them.

When new user needs emerge, that prompt the consideration of a new app, we want to consider the following:

> - Is this need a variation on one our existing applications meet? If so, improve the existing app

Reviewer: I think this reads more clearly:

```suggestion
- Is this need a variation of something that one of our existing applications meet? If so, improve the existing app
```

Many GOV.UK conventions are documented. There is the [Service Manual](https://www.gov.uk/service-manual) on the high level team approach. There are organisational tech standards in the [GDS Way](https://gds-way.cloudapps.digital/). GOV.UK has [conventions for Rails applications](https://docs.publishing.service.gov.uk/manual/conventions-for-rails-applications.html) and individual repositories can have [team agreed conventions](https://github.com/alphagov/whitehall/blob/main/docs/adr/0001-add-view-components.md). 

Not all conventions are documented. Some are learned from reading the code and the early decisions the developers made; others can be learned from observing team members or through code reviews. Not all conventions are in how software is written or documented. Others are in how team members interact with each other and collaborate.

Starting new conventions is a team activity not an individual one - we want the results of our teams to feel collective. Please communicate liberally in starting a convention so it has the best chance of becoming established and adopted.

> #### Code should be correct, clear, concise and tested

Reviewer: I agree with everything in this section, and the "Embrace existing conventions; start new conventions collaboratively" and "Optimise for maintainability, yet prioritise security" sections - but they don't feel Publishing-specific. Should they be moved to a general "code standards" section of the GOV.UK Developer Docs?

Author: Yeah - I think all of these could apply broader than Publishing so we've not aimed for them to be Publishing specific, just aimed to get agreement only in Publishing (which is more relevant for the product direction one). I'll plan to keep them in here - I'd not be opposed though, should people find these useful, for them to be re-purposed/refactored into GOV.UK wide principles.

Code should be optimised for readability. Choose clarity over brevity or performance, unless there is a genuine need not to. 

Code should have automated tests. These tests should be [necessary and sufficient](https://blog.testdouble.com/posts/2020-02-25-necessary-and-sufficient/).

> Embrace relevant software engineering principles such as [separation of concerns](https://en.wikipedia.org/wiki/Separation_of_concerns), [single responsibily principle](https://en.wikipedia.org/wiki/Single-responsibility_principle), [You aren’t gonna need it (YAGNI)](https://en.wikipedia.org/wiki/You_aren't_gonna_need_it) and [The boy scout rule](https://www.oreilly.com/library/view/97-things-every/9780596809515/ch08.html). Embrace consistency, modularity and abstractions to manage complexity. Place value in [intuitive naming](https://medium.com/signal-v-noise/hunting-for-great-names-in-programming-16f624c8fc03).

Reviewer 2: Typo: "responsibily".

Speaking of intuitive naming: the name "the boy scout rule" is not particularly intuitive. Could we come up with something more intuitive and more inclusive? ("Drive-by improvements"? "Opportunistic enhancements"?)

Author: Thanks re the typo, I've fixed it.

Yeah, I was a bit uncomfortable with the gendering of Boy Scout rule and did spend some time searching for a linkable principle that was effectively a synonym and didn't find a good example. Other suggestions would be great. We could potentially just replace it as a linked principle with a sentence of "Value that when changing code, you leave it in as-good or better state than in how you found it".

Reviewer 3: I think I've heard people just talk about it as the Campsite Rule:
- https://wolfgang-ziegler.com/blog/Development-Practice-The-Campsite-Rule.aspx 
- https://medium.com/@fooyay/the-campsite-rule-60cbf1054d92
- https://cultivatingsoftware.wordpress.com/2013/09/09/the-campsite-rule/

...which is ungendered although only slightly more intuitive.

(although: I would say maybe this is an okay use of the term because we're using it as a link text to a page which is headed "The Boy Scout Rule"? I wouldn't say "single responsibility principle" or "you aren't gonna need it" are super intuitive names either, so the fact you've linked to better explanations sort of covers the fact that naming is hard).

Reviewer 2: The Campsite Rule reads better! And it allows us to link the definition to anyone other than Robert C Martin, so we don't endorse someone[1] with a long[2] history[3] of exclusionary[4] behaviour[5] in a document giving technical direction to the entire Publishing community.

[1] https://blog.wesleyac.com/posts/robert-martin
[2] https://corgibytes.com/blog/2019/02/14/why-we-changed-a-core-value/
[3] https://www.madetech.com/blog/no-uncle-of-mine/
[4] https://web.archive.org/web/20200713003355/https://techexplained.substack.com/p/tech-bullshit-explained-uncle-bob
[5] https://web.archive.org/web/20200914214901mp_/https://meaganwaller.com/framework-whipped/

Author: Thanks, yup also good shout on avoiding reference to a divisive figure.

I've gone with adding the words directly rather than referencing "the Campsite Rule" as I felt that, unlike the other terms, there wasn't likely to be developer familiarity. By googling it, I also found that the phrase "Campsite Rule" has some sexual relationship usage - albeit in a very similar context to this - that made me a little wary.

Change is in: https://github.com/alphagov/govuk-rfcs/compare/75a520efa36f5caf802c6e061efbca0eb164ddcd..9c2af0c500d17ff9bcea2b3e8064442d182f04ae

The decisions you make in producing your code should be included in your commit message. It is common that GOV.UK developers look at old commits when trying to resolve problems or understand design decisions. Always include the “why” of a change, the “what” is usually covered by the code.

#### If a migration is worth starting, it’s worth ending

Like many tech organisations, we’ve been burnt by large projects to adapt software from an old state to a new one that have been left incomplete. We value these migrations reaching an end state, whether it’s completing the planned migration or rolling back on an idea that won’t work. We discourage staying in an intermediary state.

> If you’re part of a team that has inherited an incomplete migration, we encourage you to consider likelihood of the migrations success and whether it fits into future goals and seek to bring it to a close.

Reviewer: ```suggestion
If you’re part of a team that has inherited an incomplete migration, we encourage you to consider likelihood of the migration's success and whether it fits into future goals. Seek to bring it to a close.
```

If you’re a team starting a migratory project, plan carefully. Minimise time spent in an intermediary state, by deferring anything non-essential to the migration, particularly if the migration is going to take more than a couple of weeks (most migrations take longer than expected). Be extremely careful starting a migration that may take longer than the foreseeable roadmap, you must be sure there will remain a business priority to fund it. Consider that the longer a migration is put off, the longer it’s likely to take - keep tasks like version upgrades small by doing them often.

#### Prefer identifying and implementing missing features over toil tasks and data manipulation

Treat the need for developers to intervene in the running of an application as an efficiency smell, whether through doing things for users, running manual tasks or writing code to manipulate data. 

> Use the evidence of these tasks to understand whether applications are missing features or has resolvable limitations and, where appropriate, implement changes to reduce the need for these tasks.

Author: Thank you, fixed

> Avoid letting developer interventions blur the boundary of the service offered to users. Consider each time that a developer does something manually for a user as GOV.UK creating an undocumented feature that carries the expense of a developer’s time each usage. For frequent requests, that are not suitable as user features, document the boundaries of the service developers will provide and a runbook of how to achieve them.

Reviewer: Is it worth adding an example to the last sentence? We don't really do formal runbooks at GOV.UK, but I imagine there's a "common tasks" 2nd line page in the Dev Docs we could link to.

Author: That's a good shout, this might drift slightly into aspirational which wasn't the intention Al and I had. Let me see if I can find a good example of us documenting a dev manual task and update this.

#### Optimise for maintainability, yet prioritise security

Likely the greatest technological struggle we’ve experienced in GOV.UK has been our ability to maintain the quantity of applications we’ve created. There are many things we’d compromise for ease of maintainability but the area we won’t compromise is security.

To produce maintainable applications, we encourage the following practices:

- build applications using similar technologies (unless there is a genuinely distinct need)
- consistent approaches across an application
- documentation of architectural decisions, domain concepts, support tasks and app limitations
- leverage open source/common industry approaches for problems that aren’t unique to our domain
- avoid superfluous applications that perform similar functionality, preferring a larger application over multiple small applications.

#### Understand the tooling you own; respect the ownership of others

> Each GOV.UK repository should be owned by a team. GOV.UK has adopted a team based ownership approach to tools as we believe it results in better quality software, teams have a degree of autonomy in managing the improvement of software they own and the responsibility to fix defects can be easily established. Team members should know which software they own and what the responsibilities [of owning it](https://docs.publishing.service.gov.uk/manual/ownership-meaning.html) are. Owning teams should assist other teams that make use of that software or seek changes to the software and, should an owning team make changes that affect other teams, they should consult those teams.

Reviewer: ```suggestion
Each GOV.UK repository should be owned by a team. GOV.UK has adopted a team based ownership approach to tools as we believe it results in better quality software, teams have a degree of autonomy in managing the improvement of software they own and the responsibility to fix defects can be easily established. Team members should know which software they own and what the [responsibilities of owning it](https://docs.publishing.service.gov.uk/manual/ownership-meaning.html) are. Owning teams should assist other teams that make use of that software or seek changes to the software and, should an owning team make changes that affect other teams, they should consult those teams.
```

#### Paying down tech debt is a continuous process

Assume that in GOV.UK Publishing there will always be some degree of tech debt and that paying it down is never a single operation. Managing tech debt should be part of a process: a means to identify debts, an approach to prioritising paydown and how to celebrate the success of removing it.

Be careful in choosing to introduce tech debt for short term gain. The best time to paydown tech debt is just after it is introduced while it is fresh in memories and people understand the area. Debt that has been around for a long time can be much harder to justify the pay down as it easily falls into the realm of “we know it’s bad, but it’s been bad for 5 years so how can it be urgent?”.

#### Keep coding fun! Reduce developer pain points and help others make progress

> Most of us started coding because it is fun and we want it to keep that as part of working on GOV.UK Publishing. 

Reviewer: ```suggestion
Most of us started coding because it is fun and we want it to keep being that as part of working on GOV.UK Publishing. 
```
